deepset_cloud_sdk/s3/upload.py

```diff
@@ -0,0 +1,204 @@
+class S3UploadSummary:
+    """A summary of the S3 upload results."""
+
+    total_files: int
+    successful_upload_count: int
+    failed_upload_count: int
+    failed: List[str]
+
+
+@dataclass
+class S3UploadResult:
+    """Stores the result of an upload to S3."""
+
+    file_name: str
+    success: bool
+
+
+def make_safe_file_name(file_name: str) -> str:
+    """
+    Transform a given string to a representation that can be accepted by S3.
+
+    :param str: The file name
+    :return str: The transformed string
+    See for details of character exclusions:
+    https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html
+    """
+    transformed = re.sub(r"[\\\\#%\"'\|<>\{\}`\^\[\]~\x00-\x1F]", "_", file_name)
+    return quote(transformed)
+
+
+class S3:
+    """Client for S3 operations related to deepset Cloud uploads."""
+
+    def __init__(self, concurrency: int = 120):
+        """
+        Initialize the client.
+
+        :param concurrency: The number of concurrent upload requests
+        """
+        self.connector = aiohttp.TCPConnector(limit=concurrency)
+        self.semaphore = asyncio.BoundedSemaphore(concurrency)
+
+    @retry(retry=retry_if_exception_type(HTTPError), stop=stop_after_attempt(3), wait=wait_fixed(0.5))  # type: ignore
+    async def _upload_file_with_retries(
+        self,
+        file_name: str,
+        upload_session: UploadSession,
+        content: Any,
+        client_session: aiohttp.ClientSession,
+    ) -> aiohttp.ClientResponse:
+        """Upload a file to the prefixed S3 namespace.
+
+        :param file_path: The Path to upload from.
+        :param upload_session: UploadSession to associate the upload with.
+        :param client_session: The aiohttp ClientSession to use for this request.
+        :return: ClientResponse object.
+        """
+        aws_safe_name = make_safe_file_name(file_name)
+        aws_config = upload_session.aws_prefixed_request_config
+
+        file_data = aiohttp.FormData()
+        for key in aws_config.fields:
+            file_data.add_field(key, aws_config.fields[key])
+        file_data.add_field("file", content, filename=aws_safe_name, content_type="text/plain")
+        async with client_session.post(
+            aws_config.url,
+            data=file_data,
+        ) as response:
+            response.raise_for_status()
+
+        return response
+
+    async def upload_from_file(
+        self,
+        file_path: Path,
+        upload_session: UploadSession,
+        client_session: aiohttp.ClientSession,
+    ) -> S3UploadResult:
+        """Upload a file to the prefixed S3 namespace given a path.
+
+        :param file_path: The Path to upload from.
+        :param upload_session: UploadSession to associate the upload with.
+        :param client_session: The aiohttp ClientSession to use for this request.
+        :return: S3UploadResult object.
+        """
+        async with self.semaphore:
+            with open(file_path, "rb") as file:
```

Comment:
with open(...) is a sync operation. This will make the whole implementation sync again. We can use aiofiles for this like we did in the deepset-toolkit [here](https://github.com/deepset-ai/deepset-toolkit/blob/8641a668570c23acabaa8b57782e766baf1f0a44/deepset_toolkit/dc/asyncio_upload.py#L101) (internally)

```
import asyncio
import aiofiles

async def read_file():
    async with aiofiles.open('file.txt', mode='r') as file:
        contents = await file.read()
        print(contents)

asyncio.run(read_file())
```

Reply:
interesting, maybe that will add some additional speed, but based on evidence from the tool we created before we still got very high throuput using this